geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -116,142 +139,180 @@ object ScalaSimpleFeature {
+      this(sft, values)
+      this.id = id
+      this.userData = Collections.unmodifiableMap(userData)
+    }
 
     override def getAttribute(index: Int): AnyRef = values(index)
+    override def getUserData: util.Map[AnyRef, AnyRef] = userData
+
+    override def calculateSizeOf(): Long = Sizable.sizeOf(this) + Sizable.deepSizeOf(id, values, userData)
   }
 
   /**
     * Lazily evaluated, immutable simple feature implementation
     *
     * @param sft simple feature type
-    * @param id simple feature id
-    * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
-    * @param userDataReader lazily read user data, must be thread-safe
     */
-  class LazyImmutableSimpleFeature(
-      sft: SimpleFeatureType,
-      id: String,
-      private var reader: LazyAttributeReader,
-      private var userDataReader: LazyUserDataReader
-    ) extends AbstractImmutableSimpleFeature(sft, id) {
+  class LazyImmutableSimpleFeature private (sft: SimpleFeatureType)
+      extends AbstractImmutableSimpleFeature(sft) with LazySimpleFeature with Sizable {
 
-    // we synchronize on the low-level words, in order to minimize contention
-    private val bits = WordBitSet(sft.getAttributeCount)
-    private lazy val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
-
-    private var phaser: Phaser = new Phaser(sft.getAttributeCount) {
-      override def onAdvance(phase: Int, registeredParties: Int): Boolean = {
-        // once all attributes have been read, dereference any backing resources so that they can be gc'd
-        reader = null
-        phaser = null
-        true
-      }
+    /**
+     * Constructor
+     *
+     * @param sft simple feature type
+     * @param id simple feature id
+     * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
+     * @param userDataReader lazily read user data, must be thread-safe
+     *
+     */
+    def this(sft: SimpleFeatureType, id: String, reader: LazyAttributeReader, userDataReader: LazyUserDataReader) = {
+      this(sft)
+      this.id = id
+      this.reader = reader
+      this.userDataReader = new ImmutableLazyUserDataReader(userDataReader)
     }
 
-    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] = {
-      val read = userDataReader.read()
-      userDataReader = null // dereference any backing resources
-      Collections.unmodifiableMap(read)
-    }
-
-    override def getAttribute(index: Int): AnyRef = {
-      if (reader != null) {
-        val word = bits.word(index)
-        word.synchronized {
-          if (word.add(index)) {
-            attributes(index) = reader.read(index)
-            phaser.arriveAndDeregister()
-          }
-        }
-      }
-      attributes(index)
-    }
+    override def calculateSizeOf(): Long =
+      Sizable.sizeOf(this) + Sizable.deepSizeOf(id, bits, attributes, userData, count, reader, userDataReader)
```

Comment:
Hmm... is the reader specific to this SimpleFeature?  If so, I suppose counting its size makes sense.

Otherwise, can we share a reader?

Reply:
it is specific to this feature, as it contains the backing data (e.g. byte array) and offset into it for this feature.